File: DSA/Anisha_Parveen/Day_9_Ques_2.cpp
#include <bits/stdc++.h>
using namespace std;

bool isBalanced(const string& s) {
    stack<char> stk;
    for (char ch : s) {
        if (ch == '(' || ch == '{' || ch == '[') {
            stk.push(ch);
        } else {
            if (stk.empty()) {
                return false;
            }
            char top = stk.top();
            if ((ch == ')' && top == '(') ||
                (ch == '}' && top == '{') ||
                (ch == ']' && top == '[')) {
                stk.pop();
            } else {
                return false;
            }
        }
    }
    return stk.empty();
}

int main() {
    string s = "{[()]}";
    if (isBalanced(s)) {
        cout << "YES" << endl;
    } else {
        cout << "NO" << endl;
    }
    return 0;
}

File: DSA/Ananya_Gambhir/Day9_ques2.java
import java.util.Stack;

public class Day9_ques2 {

    public static String isBalanced(String brackets) {
        Stack<Character> stack = new Stack<>();
        
        for (char ch : brackets.toCharArray()) {
            
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
            }
            
            else if (ch == ')' || ch == '}' || ch == ']') {
                if (stack.isEmpty()) {
                    return "NO";  
                }
                char top = stack.pop();
                if (!isMatchingPair(top, ch)) {
                    return "NO"; 
                }
            }
        }
        return stack.isEmpty() ? "YES" : "NO";
    }

    private static boolean isMatchingPair(char open, char close) {
        return (open == '(' && close == ')') ||
               (open == '{' && close == '}') ||
               (open == '[' && close == ']');
    }

    public static void main(String[] args) {
        String string = "{[()]}";
        System.out.println(isBalanced(string));  
    }
}
